File: 152. Maximum Product Subarray/main.py
import sys
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_product = -sys.maxsize - 1
        left_product = 1
        right_product = 1
        size = len(nums)

        for i in range(0, size):
            left_product *= nums[i]
            right_product *= nums[size - 1 - i]
            max_product = max(max_product, left_product)
            max_product = max(max_product, right_product)
            if left_product == 0:
                left_product = 1
            if right_product == 0:
                right_product = 1

        return max_product


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_product = -sys.maxsize - 1
        cur_max = 1
        cur_min = 1

        for num in nums:
            temp = cur_max * num
            cur_max = max(temp, cur_min * num, num)
            cur_min = min(temp, cur_min * num, num)
            max_product = max(max_product, cur_max)

        return max_product

File: 152. Maximum Product Subarray/main.rs
struct Solution;

impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut max_prod = i32::MIN;
        let mut left_prod = 1;
        let mut right_prod = 1;
        let size = nums.len();

        for i in 0..size {
            left_prod *= nums[i];
            right_prod *= nums[size - 1 - i];
            max_prod = max_prod.max(left_prod);
            max_prod = max_prod.max(right_prod);
            if left_prod == 0 {
                left_prod = 1;
            }
            if right_prod == 0 {
                right_prod = 1;
            }
        }

        max_prod
    }
}

impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let mut max_prod = i32::MIN;
        let mut cur_min = 1;
        let mut cur_max = 1;

        for num in nums {
            let temp = cur_max * num;
            cur_max = temp.max(cur_min * num).max(num);
            cur_min = temp.min(cur_min * num).min(num);
            max_prod = max_prod.max(cur_max);
        }

        max_prod
    }
}
